### PR TITLE
Refactor well-known types

### PR DESCRIPTION
**Pull request by sharwell**
Recommended review strategy is commit-by-commit.

### CODE REVIEW

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs [hunk 1]
```diff
@@ -33,43 +32,26 @@ internal static void Main(SuppressibleLoggingHelper? log, string referenceAssemb
...
+            attributesOfInterest.Add(wellKnownTypes.SystemDiagnosticsCodeAnalysisMaybeNullAttribute.Value.FullName, wellKnownTypes.SystemDiagnosticsCodeAnalysisMaybeNullAttribute.Value.Resolve());
+            attributesOfInterest.Add(wellKnownTypes.SystemDiagnosticsCodeAnalysisMaybeNullWhenAttribute.Value.FullName, wellKnownTypes.SystemDiagnosticsCodeAnalysisMaybeNullWhenAttribute.Value.Resolve());
+            attributesOfInterest.Add(wellKnownTypes.SystemDiagnosticsCodeAnalysisNotNullAttribute.Value.FullName, wellKnownTypes.SystemDiagnosticsCodeAnalysisNotNullAttribute.Value.Resolve());
+            attributesOfInterest.Add(wellKnownTypes.SystemDiagnosticsCodeAnalysisNotNullIfNotNullAttribute.Value.FullName, wellKnownTypes.SystemDiagnosticsCodeAnalysisNotNullIfNotNullAttribute.Value.Resolve());
+            attributesOfInterest.Add(wellKnownTypes.SystemDiagnosticsCodeAnalysisNotNullWhenAttribute.Value.FullName, wellKnownTypes.SystemDiagnosticsCodeAnalysisNotNullWhenAttribute.Value.Resolve());
```
Comment (jnm2) [inline TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs]: What do you think of:

```cs
foreach (var attribute in new[]
{
    wellKnownTypes.SystemRuntimeCompilerServicesNullableAttribute,
    wellKnownTypes.SystemRuntimeCompilerServicesNullableContextAttribute,
    ...
})
{
    attributesOfInterest.Add(attribute.Value.FullName, attribute.Resolve());
}
```

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+MaybeNullWhenAttributeProvidedType.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            {
+            }
+
+            protected override TypeReference DefineAttribute(ModuleDefinition module, WellKnownTypes wellKnownTypes, CustomAttributeFactory attributeFactory)
+            {
+                var attribute = new TypeDefinition(
+                    @namespace: NamespaceName,
+                    name: TypeName,
+                    TypeAttributes.NotPublic | TypeAttributes.Sealed | TypeAttributes.BeforeFieldInit,
+                    wellKnownTypes.Module.ImportReference(wellKnownTypes.SystemAttribute));
```
Comment (jnm2) [inline TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+MaybeNullWhenAttributeProvidedType.cs]: Protected helper methods could cut down on some of the remaining repetition.

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+WellKnownType.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            public TypeReference GetOrCreateTypeReference(ModuleDefinition module, WellKnownTypes wellKnownTypes)
+            {
+                if (_typeReference is null)
+                {
+                    lock (_syncObject)
+                    {
+                        if (_typeReference is null)
+                        {
+                            _typeReference = GetOrCreateTypeReferenceImpl(module, wellKnownTypes);
+                        }
```
Comment (jnm2) [inline TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+WellKnownType.cs]: C# 8 (if desired):

```cs
_typeReference ??= GetOrCreateTypeReferenceImpl(module, wellKnownTypes);
```

**Review by jnm2 — COMMENTED** [review]
Very clean!

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs (hunk 1 above)
Comment (sharwell) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs]: I was actually thinking of just extracting a local function to add the items.

Comment (jnm2) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs]: That works too! Or foreach over `wellKnownTypes.AllNullabilityAttributes`.

Comment (sharwell) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs]: Now simplified

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+WellKnownType.cs (hunk 3 above)
Comment (sharwell) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+WellKnownType.cs]: Nice. Now implemented.

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs (hunk 1 above)
Comment (jnm2) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/Program.cs]: Or `wellKnownTypes.AllNullabilityAttributes.ToDictionary(a => a.Value.FullName, a.Value.Resolve());`

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+MaybeNullWhenAttributeProvidedType.cs (hunk 2 above)
Comment (sharwell) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+MaybeNullWhenAttributeProvidedType.cs]: Yeah that works nicely. Now implemented.

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+ProvidedType.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+            public string TypeName { get; }
+
+            public string FullTypeName { get; }
+
+            protected override TypeReference GetOrCreateTypeReferenceImpl(ModuleDefinition module, WellKnownTypes wellKnownTypes)
+            {
+                var existingType = module.TypeSystem.Object.Resolve().Module.GetType(FullTypeName);
+                if (existingType is object && existingType.IsPublic)
+                    return existingType;
```
Comment (jnm2) [inline TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+ProvidedType.cs:29]: Could be

```cs
var existingType = module.TypeSystem.Object.Resolve().Module.GetType(FullTypeName);
if (existingType is { IsPublic: true })
    return existingType;
```

Or

```cs
if (module.TypeSystem.Object.Resolve().Module.GetType(FullTypeName) is { IsPublic: true } existingType)
    return existingType;
```

(matter of taste for sure, I think Cyrus would hate the last one)

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+NullableAttributeProvidedType.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            internal NullableAttributeProvidedType()
+                : base("System.Runtime.CompilerServices", "NullableAttribute")
+            {
+            }
+
+            protected override void ImplementAttribute(ModuleDefinition module, TypeDefinition attribute, WellKnownTypes wellKnownTypes, CustomAttributeFactory attributeFactory)
+            {
+                MethodDefinition compilerGeneratedConstructor = wellKnownTypes.SystemRuntimeCompilerServicesCompilerGeneratedAttribute.Resolve().Methods.Single(method => method.IsConstructor && !method.IsStatic && method.Parameters.Count == 0);
+                attribute.CustomAttributes.Add(new CustomAttribute(wellKnownTypes.Module.ImportReference(compilerGeneratedConstructor)));
+                attribute.CustomAttributes.Add(new CustomAttribute(wellKnownTypes.MicrosoftCodeAnalysisEmbeddedAttribute.Value.Resolve().Methods.Single(method => method.IsConstructor && !method.IsStatic && method.Parameters.Count == 0)));
```
Comment (jnm2) [inline TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+NullableAttributeProvidedType.cs]: Could add a protected helper method to add the two attributes that keep showing up.

**Review by jnm2 — APPROVED** [review]

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+NullableAttributeProvidedType.cs (hunk 5 above)
Comment (jnm2) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+NullableAttributeProvidedType.cs]: (And maybe the method name could help explain why they are being added)

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+ProvidedType.cs (hunk 4 above)
Comment (sharwell) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+ProvidedType.cs:29]: I'm good with the current form here

File: TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+NullableAttributeProvidedType.cs (hunk 5 above)
Comment (sharwell) [reply TunnelVisionLabs.ReferenceAssemblyAnnotator/WellKnownTypes+NullableAttributeProvidedType.cs]: Updated the code to use existing `CustomAttributeFactory` methods instead.

**Pull request merged**